### PR TITLE
Exec plugin integration test

### PR DESCRIPTION
**Pull request by ankeesler**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

- This PR adds more integration tests for the client-go credential provider feature.
- We are taking the feature GA, and thus need to implement the [test plan from the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/541-external-credential-providers/README.md#test-plan).
- KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/541-external-credential-providers/README.md.

**Which issue(s) this PR fixes**:

- Feature set tracking issue: https://github.com/kubernetes/enhancements/issues/541
- KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/541-external-credential-providers/README.md

**Special notes for your reviewer**:

- There are more integration tests to write, but I am trying to work in smaller iterations.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: test/integration/client/exec_test.go [hunk 1]
```diff
@@ -464,3 +568,61 @@ func loadX509KeyPair(certFile, keyFile string) *tls.Certificate {
...
+	})
+
+	return ns
+}
+
+func startConfigMapInformer(ctx context.Context, t *testing.T, client clientset.Interface, namespace string) *informerSpy {
+	t.Helper()
+
+	var informerSpy informerSpy
+	informers := informers.NewSharedInformerFactoryWithOptions(client, 0, informers.WithNamespace(namespace))
```
Comment (enj) [inline test/integration/client/exec_test.go:599]: nit: rename `informers` to not conflict with the import name.

File: test/integration/client/exec_test.go [hunk 2]
```diff
@@ -464,3 +568,61 @@ func loadX509KeyPair(certFile, keyFile string) *tls.Certificate {
...
+	informers.Core().V1().ConfigMaps().Informer().AddEventHandler(&informerSpy)
+	informers.Start(ctx.Done())
+	synced := informers.WaitForCacheSync(ctx.Done())
```
Comment (enj) [inline test/integration/client/exec_test.go:602]: nit: assert `synced` has length `1`

File: test/integration/client/exec_test.go [hunk 3]
```diff
@@ -407,6 +382,135 @@ func TestExecPlugin(t *testing.T) {
...
+			// Validate that the informer was called correctly.
+			if want, got := 1, len(informerSpy.adds); want != got {
+				t.Errorf("wanted %d add event(s), got %d (%s)", want, got, informerSpy.adds)
+			}
+			if want, got := 1, len(informerSpy.updates); want != got {
+				t.Errorf("wanted %d update event(s), got %d (%s)", want, got, informerSpy.updates)
+			}
+			if want, got := 1, len(informerSpy.deletes); want != got {
+				t.Errorf("wanted %d add event(s), got %d (%s)", want, got, informerSpy.deletes)
+			}
```
Comment (enj) [inline test/integration/client/exec_test.go:480]: Assert that the events match what you expect to see.

File: test/integration/client/exec_test.go [hunk 4]
```diff
@@ -407,6 +382,135 @@ func TestExecPlugin(t *testing.T) {
...
+
+func (es *informerSpy) OnAdd(obj interface{}) { es.adds = append(es.adds, obj) }
+
+func (es *informerSpy) OnUpdate(old, new interface{}) {
+	es.updates = append(es.updates, oldNew{old: old, new: new})
+}
+
+func (es *informerSpy) OnDelete(obj interface{}) { es.deletes = append(es.deletes, obj) }
+
+func TestExecPluginViaInformer(t *testing.T) {
```
Comment (enj) [inline test/integration/client/exec_test.go:403]: While I think the existing `2` tests are fine as a starting point, we really need to test the informer over time - i.e., if the creds under it rotate, does it keep working (and if the creds stop working for some time, and then start working again, does it recover, ...).

**Review by enj — CHANGES_REQUESTED** [review]
Minor stuff.

Reminder that code thaw is April 8, 2021.

File: test/integration/client/exec_test.go (hunk 3 above)
Comment (ankeesler) [reply test/integration/client/exec_test.go]: e2105d431ee

File: test/integration/client/exec_test.go (hunk 4 above)
Comment (ankeesler) [reply test/integration/client/exec_test.go:408]: Totes agreed. That is the next step. Figured I would start small to follow a "simple PR first" principle.

File: test/integration/client/exec_test.go (hunk 2 above)
Comment (ankeesler) [reply test/integration/client/exec_test.go]: d7c6feb7273

File: test/integration/client/exec_test.go (hunk 1 above)
Comment (ankeesler) [reply test/integration/client/exec_test.go]: 9a64b4195ab

**Comment by enj** [comment]
/milestone v1.22

??

**Review by enj — COMMENTED** [review]
Just some last nits, go ahead and squash.

**Comment by enj** [comment]
/lgtm
/approve
/retest

This should automatically merge once the 1.22 milestone opens on the master branch.

**Pull request merged**